File: src/main/java/com/ptaf/ai/validation/AllowedYamlGuard.java
package com.ptaf.ai.validation;

import com.ptaf.ai.index.YamlKeyIndex;
import com.ptaf.ai.model.AiGenerationStructuredResponse;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class AllowedYamlGuard {
    private static final Pattern YAML_KEY_PATTERN =
            Pattern.compile("(elements|api_requests|queries|config)\\.[a-zA-Z0-9_.-]+");

    public AllowedYamlGuardResult validate(
            AiGenerationStructuredResponse structuredResponse,
            YamlKeyIndex yamlKeyIndex
    ) {
        Set<String> known = yamlKeyIndex.normalizedKeys();
        Set<String> used = normalizeList(structuredResponse.yamlKeysUsed());
        Set<String> missingDeclared = normalizeList(structuredResponse.missingYamlKeys());
        Set<String> keysInFeature = extractYamlLikeKeys(structuredResponse.featureFile());

        List<String> allowed = new ArrayList<>();
        List<String> unknown = new ArrayList<>();
        List<String> warnings = new ArrayList<>();
        List<String> blocking = new ArrayList<>();
        List<String> missingUsedInFeature = new ArrayList<>();

        for (String key : used) {
            if (known.contains(key)) {
                allowed.add(key);
            } else {
                unknown.add(key);
                blocking.add("YAML key does not exist: " + key);
            }
        }

        for (String key : missingDeclared) {
            if (keysInFeature.contains(key)) {
                missingUsedInFeature.add(key);
                blocking.add("Missing YAML key cannot be used directly in FEATURE_FILE: " + key);
            }
        }

        for (String key : keysInFeature) {
            boolean isKnown = known.contains(key);
            boolean listedUsed = used.contains(key);
            if (!isKnown) {
                if (!unknown.contains(key)) {
                    unknown.add(key);
                }
                blocking.add("Unknown YAML-looking key used in FEATURE_FILE: " + key);
            } else if (!listedUsed) {
                warnings.add("Known YAML key appears in FEATURE_FILE but not listed in YAML_KEYS_USED: " + key);
            }
        }

        return new AllowedYamlGuardResult(
                blocking.isEmpty(),
                allowed,
                unknown,
                new ArrayList<>(missingDeclared),
                missingUsedInFeature,
                dedupe(warnings),
                dedupe(blocking)
        );
    }

    private static Set<String> normalizeList(List<String> keys) {
        Set<String> out = new LinkedHashSet<>();
        if (keys == null) {
            return out;
        }
        for (String key : keys) {
            String normalized = YamlKeyIndex.normalizeKey(key);
            if (!normalized.isBlank()) {
                out.add(normalized);
            }
        }
        return out;
    }

    private static Set<String> extractYamlLikeKeys(String featureFile) {
        Set<String> out = new LinkedHashSet<>();
        if (featureFile == null || featureFile.isBlank()) {
            return out;
        }
        Matcher matcher = YAML_KEY_PATTERN.matcher(featureFile);
        while (matcher.find()) {
            String key = YamlKeyIndex.normalizeKey(matcher.group());
            if (!key.isBlank()) {
                out.add(key);
            }
        }
        return out;
    }

    private static List<String> dedupe(List<String> values) {
        return new ArrayList<>(new LinkedHashSet<>(values));
    }
}

File: ptaf/ai/validation/allowed_yaml_guard.py
from __future__ import annotations

import re

from ptaf.ai.index.yaml_key_index import YamlKeyIndex
from ptaf.ai.model.ai_generation_structured_response import AiGenerationStructuredResponse
from ptaf.ai.validation.allowed_yaml_guard_result import AllowedYamlGuardResult

_YAML_KEY_PATTERN = re.compile(r"(elements|api_requests|queries|config)\.[a-zA-Z0-9_.-]+")


class AllowedYamlGuard:
    def validate(
        self,
        structured_response: AiGenerationStructuredResponse,
        yaml_key_index: YamlKeyIndex,
    ) -> AllowedYamlGuardResult:
        known = yaml_key_index.normalized_keys()
        used = _normalize_list(structured_response.yaml_keys_used)
        missing_declared = _normalize_list(structured_response.missing_yaml_keys)
        keys_in_feature = _extract_yaml_like_keys(structured_response.feature_file)

        allowed: list[str] = []
        unknown: list[str] = []
        warnings: list[str] = []
        blocking: list[str] = []
        missing_used_in_feature: list[str] = []

        for key in used:
            if key in known:
                allowed.append(key)
            else:
                unknown.append(key)
                blocking.append(f"YAML key does not exist: {key}")

        for key in missing_declared:
            if key in keys_in_feature:
                missing_used_in_feature.append(key)
                blocking.append(f"Missing YAML key cannot be used directly in FEATURE_FILE: {key}")

        for key in keys_in_feature:
            is_known = key in known
            listed_used = key in used
            if not is_known:
                if key not in unknown:
                    unknown.append(key)
                blocking.append(f"Unknown YAML-looking key used in FEATURE_FILE: {key}")
            elif not listed_used:
                warnings.append(
                    f"Known YAML key appears in FEATURE_FILE but not listed in YAML_KEYS_USED: {key}"
                )

        return AllowedYamlGuardResult(
            passed=not blocking,
            allowed_keys=allowed,
            unknown_keys_used=unknown,
            missing_keys_declared=list(missing_declared),
            missing_keys_used_in_feature=missing_used_in_feature,
            warnings=_dedupe(warnings),
            blocking_errors=_dedupe(blocking),
        )


def _normalize_list(keys: list[str] | None) -> set[str]:
    out: set[str] = set()
    if not keys:
        return out
    for key in keys:
        normalized = YamlKeyIndex.normalize_key(key)
        if normalized:
            out.add(normalized)
    return out


def _extract_yaml_like_keys(feature_file: str | None) -> set[str]:
    out: set[str] = set()
    if not feature_file or not feature_file.strip():
        return out
    for match in _YAML_KEY_PATTERN.finditer(feature_file):
        key = YamlKeyIndex.normalize_key(match.group())
        if key:
            out.add(key)
    return out


def _dedupe(values: list[str]) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for value in values:
        if value not in seen:
            seen.add(value)
            out.append(value)
    return out
